Docs: Module mocking, and more

kylegach opened the pull request:
Closes #26795, closes #26796, closes #26797

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

Add or update docs for:

- Storybook for Next.js
    - Add or adjust sections for:
        - Module aliases
        - Subpath imports
        - Mocking modules
        - Portable stories
        - Exported modules API
- Decorators
    - Focus "Context" for mocking section on just the second argument
        - Move rest to Story rendering
- Build pages with Storybook
    - Simplify to reference new mocking guides
- Mocking data and modules - all new
    - Mocking providers 
    - Mocking network requests
    - Mocking modules
- Interaction testing
    - Add sections for:
        - Run code before each test
        - Mocked modules
- Story rendering
    - Add section for Running code for every story
        - Move some of "Context" for mocking examples and snippets here
- Parameters API reference
    - Add `test` properties

## Checklist for Contributors

### Testing

#### Manual testing

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `module-mocking-docs`

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
WIP

kylegach committed:
Split into 3 pages

kasperpeulen committed:
Try another mock naming convention

kasperpeulen committed:
Bind this

kasperpeulen committed:
Update snapshot

kasperpeulen committed:
Fix name

kasperpeulen committed:
Hide some junk of next for nwo

kasperpeulen committed:
Update code/frameworks/nextjs/src/export-mocks/navigation/index.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update code/frameworks/nextjs/src/export-mocks/navigation/index.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update code/frameworks/nextjs/src/export-mocks/navigation/index.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Update code/frameworks/nextjs/src/export-mocks/navigation/index.ts

Co-authored-by: Yann Braga <yannbf@gmail.com>

kasperpeulen committed:
Hide some junk of next for now

JReinhold committed:
Add Modules section to Next.js docs

kasperpeulen committed:
Show redirect and next/cache spies

kylegach committed:
More mocking content

- Providers
    - Configuring mocked providers with parameters
- Network requests
    - Bad link
- Modules
    - Better mock files section
    - Using mocked files in stories

kylegach committed:
Updates for Next.js

- Warning for absolute imports & mocking
- Module aliases
- Subpath imports
- Mocking modules
- `@storybook/nextjs/cache.mock` module API

kylegach committed:
Simplify Building pages with Storybook guide

kylegach committed:
Update Decorators and Interaction tests pages

- Decorators
    - Focus the "context" section on only that argument
    - Move irrelevant examples and snippets to the Story rendering page
- Interaction Testing
    - Under the Write an interaction test section, add:
        - Run code before each test section
        - Mocked modules section

kylegach committed:
Add missing TODOs

kasperpeulen committed:
Add new actions to the bottom of the panel and auto scroll

JReinhold reviewed (commented):
Fantastic job. ❤️

JReinhold committed:
add cache.mock entrypoint to nextjs

kasperpeulen commented on docs/get-started/nextjs.md:
```diff
@@ -671,6 +671,106 @@ import 'styles/globals.scss';
+// package.json
+{
+  "imports": {
+    "#components": "./components",
```
you only need `"#*": ["./*", "./*.ts", "./*.tsx"]`

kylegach replied:
@JReinhold — Can you update this?

kasperpeulen commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -0,0 +1,229 @@
+
+- It should re-export all exports from the original module.
+- It should use the `fn` utility to mock any necessary functionality from the original module.
+- It cannot use subpath imports or builder aliases for its dependencies.
```
I would say, it should import the `actual` file as a relative import.

JReinhold replied:
Yes I think the main thing here is that you can't use subpath imports/aliases to import the _actual_ file, as that would alias to itself, causing a cyclical crash. but any other dependencies or imports can work fine however you want.

kylegach replied:
@JReinhold — Can you update this, please?

kasperpeulen commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -0,0 +1,229 @@
+      "storybook": "./api.mock.ts",
+      "default": "./api.ts"
+    },
+    "#prisma/prisma": {
```
I would use this:

```
    "#app/actions": {
      "storybook": "./app/actions.mock.ts",
      "default": "./app/actions.ts"
    },
    "#lib/session": {
      "storybook": "./lib/session.mock.ts",
      "default": "./lib/session.ts"
    },
    "#lib/db": {
      "storybook": "./lib/db.mock.ts",
      "default": "./lib/db.ts"
    },
```

kylegach replied:
@JReinhold — Can you update this?

kasperpeulen committed:
Address review

kylegach commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -0,0 +1,229 @@
+
+To configure subpath imports, you define the `imports` property in your project's `package.json` file. This property maps the subpath to the actual file path. The example below configures subpath imports for the `api` and `prisma` internal modules:
+
+TK: External module example?
```
@JReinhold — Want to take a first crack at that?

JReinhold replied:
Added some words and an example. I'm not sure the example is necessary as it's very similar to the one above. I also don't fully love it because users today rarely use `'uuid'` but instead `crypto.randomUUID()`, but that wouldn't really be an example of mocking an external module..

kylegach committed:
Add vite example

Co-authored-by: Jeppe Reinhold <jeppe@chromatic.com>

kylegach commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -0,0 +1,229 @@
+const meta: Meta<typeof Page> = {
+  component: Page,
+  async beforeEach() {
+    // 👇 Do this for each story
+    // TK
+    // 👇 Clear the mock between stories
+    getUserFromSession.mockClear();
```
@JReinhold — I welcome anything you can do to make this more accurate/clear/useful.

JReinhold replied:
Added more docs on `beforeEach` and the cleanup function.

kylegach commented on docs/writing-tests/interaction-testing.md:
```diff
@@ -90,6 +90,38 @@ Once the story loads in the UI, it simulates the user's behavior and verifies th
+### Run code before each test
+
+It can be helpful to run code before each test to set up the initial state of the component or reset the state of modules. You can do this by adding an `async beforeEach` function to the meta in your stories file. This function will run before each test in the story file.
+
```
@JReinhold — Mind adding an example snippet for that use case? That would help me understand more to be able to put in prose.

JReinhold replied:
I added stuff to `mocking-modules.md`, but now I'm unsure if it should have been here instead?!
Please cleanup my mess.

kylegach replied:
Until we have a dedicated `beforeEach` page to reference (which I'm _not_ suggesting we do now), I think duplicating the info in both places is the way to go.

JReinhold committed:
clarify mock->original import limitations.

JReinhold committed:
add paragraph on mocking external modules.

JReinhold committed:
add docs on `beforeEach`, `cleanup`, `parameters.test`

JReinhold committed:
simplify subpath example in nextjs

kylegach commented on docs/get-started/nextjs.md:
```diff
@@ -671,6 +671,106 @@ import 'styles/globals.scss';
+
+## Subpath imports
+
+As an alternative to [module aliases](#module-aliases), you can use [subpath imports](https://nodejs.org/api/packages.html#subpath-imports) to import modules. This follows Node package standards and has benefits when [mocking modules](#mocking-modules).
```
The term `subpath imports` is already linked to that exact URL. I could be misunderstanding your feedback. But otherwise, the fact that two of you missed that is noteworthy so I thought for a while about how to improve this... and came up empty. I really want to keep the _term_ linked, and long link text doesn't work well.

JReinhold replied:
I think I was too focused on the "This follows Node package standards" part. It's probably fine.

kylegach commented on docs/writing-stories/build-pages-with-storybook.md:
```diff
@@ -73,226 +73,19 @@ This approach is beneficial when the various subcomponents export a complex list
-### Mocking API Services
+### [Mocking API Services](./mocking-network-requests.md)
 
-Connected applications such as Twitter, Instagram, amongst others, are everywhere, consuming data from REST or GraphQL endpoints. Suppose you're working in an application that relies on either of these data providers. In that case, you can add Mock Service Worker (MSW) via [Storybook's MSW addon](https://storybook.js.org/addons/msw-storybook-addon) to mock data alongside your app and stories.
+For components that make network requests (e.g. fetching data from a REST or GraphQL API), you can mock those requests in your stories.
 
-[Mock Service Worker](https://mswjs.io/) is an API mocking library. It relies on service workers to capture network requests and provides mocked data in response. The MSW addon adds this functionality into Storybook, allowing you to mock API requests in your stories. Below is an overview of how to set up and use the addon.
+### [Mocking imports](./mocking-modules.md)
```
I actually did this on purpose, so that any links to these headings will still be useful.

Longer-term, I see this page going away entirely.

kylegach commented on docs/get-started/nextjs.md:
```diff
@@ -885,54 +985,59 @@ You can refer to the [Install `sharp` to Use Built-In Image Optimization](https:
+Type: `{ getPackageAliases: ({ useESM?: boolean }) => void }`
 
-Default: `false`
+`getPackageAliases` is a helper to generate the aliases needed to set up [portable stories in a Jest environment](../api/portable-stories-jest.md).
```
Good call. I added the structure, but would appreciate if you could fill the section out, if necessary.

JReinhold replied:
Added the reasoning behind it but couldn't come up with anything else that was useful and not already covered by the API docs.

kylegach committed:
Address feedback

- Next.js
    - Add portable stories section
- Mocking modules
    - Clarify requirements of mock files
    - Prose and snippet tweaks
- Interaction testing
    - Bring over `mockdate` example
    - Prose and snippet tweaks

kylegach commented on docs/get-started/nextjs.md:
```diff
@@ -885,54 +989,59 @@ You can refer to the [Install `sharp` to Use Built-In Image Optimization](https:
+`getPackageAliases` is a helper to generate the aliases needed to set up [portable stories](#portable-stories).
 
-If your story imports components that use `next/navigation`, you need to set the parameter `nextjs.appDirectory` to `true`. Because this is a parameter, you can apply it to a [single story](../api/parameters.md#story-parameters), [all stories for a component](../api/parameters.md#meta-parameters), or [every story in your Storybook](../api/parameters.md#project-parameters). See [Next.js Navigation ...
+TK: Example snippet
```
@JReinhold — I could use your help with the `TK`s in this section.

kylegach commented on docs/api/parameters.md:
```diff
@@ -141,6 +141,51 @@ When specifying a custom sorting function, the function behaves like a typical J
+
+Default: `false`
+
+Setting this to `true` will prevent the [play function](../writing-stories/play-function.md) from failing and showing a warning when an uncaught exception is thrown.
```
@JReinhold — How's this? It'd be nice to provide a common use case?

JReinhold committed:
add getPackageAliases example

JReinhold commented on docs/get-started/nextjs.md:
````diff
@@ -885,54 +989,83 @@ You can refer to the [Install `sharp` to Use Built-In Image Optimization](https:
 
-If your story imports components that use `next/navigation`, you need to set the parameter `nextjs.appDirectory` to `true`. Because this is a parameter, you can apply it to a [single story](../api/parameters.md#story-parameters), [all stories for a component](../api/parameters.md#meta-parameters), or [every story in your Storybook](../api/parameters.md#project-parameters). See [Next.js Navigation ...
+```ts
+import type { Config } from 'jest';
+import nextJest from 'next/jest.js';
+// 👇 import the utility function
+import { getPackageAliases } from '@storybook/nextjs/export-mocks';
+
+const createJestConfig = nextJest({
+  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
+  dir: './',
+});
+
+// Add any custom config to be passed to Jest
+const config: Config = {
+  coverageProvider: 'v8',
+  testEnvironment: 'jsdom',
+  // Add more setup options before each test is run
+  // setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
+  moduleNameMapper: {
+    ...getPackageAliases(), // 👈 add the utility as a moduleNameMapper
+  },
+};
 
-#### `navigation`
+// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
+export default createJestConfig(config);
+```
````
This is just [the example from the Next.js docs](https://nextjs.org/docs/app/building-your-application/testing/jest#manual-setup) plus the utility function. So it can be condensed a lot of we want to.

JReinhold committed:
add examples to next mock docs

JReinhold committed:
improve documentation for parameters.test.dangerouslyIgnoreUnhandledErrors

JReinhold committed:
remove docs on default router and navigation contexts

JReinhold commented on docs/get-started/nextjs.md:
````diff
@@ -301,41 +301,6 @@ The default values on the stubbed router are as follows (see [globals](../essent
 ```ts
 // Default router
 const defaultRouter = {
-  push(...args) {
````
The new recommended way to "override" these is with `mockImplementation`, in which case you don't really need to know about their default behavior (which I also think is different now) and the caveat about them missing from the actions panel is now also not true anymore.

kylegach replied:
Oh, great catch! 🙏

kylegach replied:
So how do you actually override one of the methods now?

```ts
// .storybook/preview.ts
import { Preview } from '@storybook/react';

const preview: Preview = {
  parameters: {
    nextjs: {
      router: {
        push: // TK: How?,
      },
    },
  },
};
```

Or do you have to use `beforeEach`?

```ts
// .storybook/preview.ts
import { Preview } from '@storybook/react';

const preview: Preview = {
  async beforeEach({ parameters }) {
    parameters.nextjs.router.push.mockImplentation(() => { /* ... */ });
  },
};
```

kylegach replied:
Oh, wait. Maybe you import the mock?

```
// .storybook/preview.ts
import { Preview } from '@storybook/react';
import { getRouter } from '@storybook/nextjs/router';

const preview: Preview = {
  parameters: {
    nextjs: {
      router: {
        push: getRouter().push.mockImplentation(() => { /* ... */ }),
      },
    },
  },
};
```

JReinhold replied:
I guess the latter could work, but the recommended way would now be this:

```ts
// .storybook/preview.ts
import { Preview } from '@storybook/react';
import { getRouter } from '@storybook/nextjs/router.mock';

const preview: Preview = {
  async beforeEach() {
    getRouter().push.mockImplementation(() => { /* ... */ });
  },
};
```

JReinhold replied:
The functions are really best modified with `mockImplementation` and the parameters should rarely be used, except for the non-function properties like `pathname`, `query`, etc. which I've also left in.

kylegach replied:
I think I found a format that demonstrates both clearly. But feel free to tweak it or make further suggestions.

JReinhold replied:
LGTM

JReinhold committed:
add reasons for aliases in nextjs+jest

kylegach committed:
Add back nextjs router/navigation override instructions

Other improvements:
- Consistency improvements in snippets throughout
- Prose tweaks

kylegach committed:
Improve MSW snippets

- Use MSW 2.0 API
- Use loader instead of decorator
- Improve snippet filename format

yannbf commented on docs/snippets/angular/msw-addon-configure-handlers-graphql.ts.mdx:
```diff
@@ -58,28 +58,38 @@ type Story = StoryObj<DocumentScreen>;
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', () => {
+          return new HttpResponse.json({
```
Don't forget to add `HttpResponse` to the msw imports at line 11

yannbf commented on docs/snippets/angular/msw-addon-configure-handlers-graphql.ts.mdx:
```diff
@@ -58,28 +58,38 @@ type Story = StoryObj<DocumentScreen>;
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
this should be awaited

Also don't forget to add `delay` to the msw imports at line 11

yannbf commented on docs/snippets/angular/msw-addon-configure-handlers.ts.mdx:
```diff
@@ -41,21 +41,28 @@ const TestData = {
+      handlers: [
+        http.get('https://your-restful-endpoint', async () => {
+          await delay(800);
+          return new HttpResponse(null, {
```
Don't forget to import `HttpResponse`, `delay`, `http` and remove `rest` from the imports at line 6

yannbf commented on docs/snippets/common/msw-addon-configure-handlers-http.ts-4-9.mdx:
```diff
@@ -42,21 +42,28 @@ const TestData = {
+      handlers: [
+        http.get('https://your-restful-endpoint', async () => {
+          await delay(800);
+          return new HttpResponse(null, {
```
Don't forget to import `HttpResponse`, `delay`, `http` and remove `rest` from the imports at line 7

yannbf commented on docs/snippets/common/msw-addon-configure-handlers-http.ts.mdx:
```diff
@@ -42,21 +42,28 @@ const TestData = {
+      handlers: [
+        http.get('https://your-restful-endpoint', async () => {
+          await delay(800);
+          return new HttpResponse(null, {
```
Don't forget to import `HttpResponse`, `delay`, `http` and remove `rest` from the imports at line 7

yannbf commented on docs/snippets/common/msw-addon-initialize.js.mdx:
```diff
@@ -0,0 +1,16 @@
+ */
+initialize();
+
+export default {
```
Not sure if this is used across the JS docs, but you could add types like so:

/**
 * @type {import('@storybook/react').Preview}
 */

kylegach replied:
That's a great idea, but should be a separate effort.

yannbf commented on docs/snippets/common/msw-addon-initialize.ts.mdx:
```diff
@@ -1,10 +1,10 @@
 // .storybook/preview.ts
 
-// Replace your-framework with the framework you are using (e.g., react, vue3)
+// Replace your-framework with the framework you are using (e.g., nextjs, vue3-vite)
 import { Preview } from '@storybook/your-framework';
```
this is a more general feedback (in case other docs have similar structure) but the `Preview` type comes from the renderer (e.g. react, vue3), while the main.js `StorybookConfig` type comes from the frameworks

yannbf commented on docs/snippets/common/msw-addon-initialize.ts.mdx:
```diff
@@ -14,15 +14,8 @@ import { initialize, mswDecorator } from 'msw-storybook-addon';
 initialize();
```
I think if you add the link to the docs in one place, it should be in all places:

```
/*
 * Initializes MSW
 * See https://github.com/mswjs/msw-storybook-addon#configuring-msw
 * to learn how to customize it
 */
```

kylegach replied:
Are you suggesting we add a similar comment above other usages of MSW?

yannbf commented on docs/snippets/react/msw-addon-configure-handlers-graphql.ts.mdx:
```diff
@@ -64,34 +64,38 @@ type Story = StoryObj<typeof SampleComponent>;
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
this should be awaited

yannbf commented on docs/snippets/svelte/msw-addon-configure-handlers-graphql.js.mdx:
```diff
@@ -37,35 +40,39 @@ const TestData = {
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
This should be awaited
Also don't forget to add `HttpResponse` and `delay` to the msw imports at line 4

yannbf commented on docs/snippets/svelte/msw-addon-configure-handlers-graphql.js.mdx:
```diff
@@ -8,6 +8,9 @@ import MockApolloWrapperClient from './MockApolloWrapperClient.svelte';
 
 export default {
   component: DocumentScreen,
+  render: () => ({
```
is this needed at all? seems like it would be equivalent to the default render function

yannbf commented on docs/snippets/svelte/msw-addon-configure-handlers-graphql.ts-4-9.mdx:
```diff
@@ -43,28 +46,38 @@ const TestData = {
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
This should be awaited
Also don't forget to add `HttpResponse` and `delay` to the msw imports at line 6

yannbf commented on docs/snippets/svelte/msw-addon-configure-handlers-graphql.ts-4-9.mdx:
```diff
@@ -10,6 +10,9 @@ import MockApolloWrapperClient from './MockApolloWrapperClient.svelte';
 
 const meta = {
   component: DocumentScreen,
+  render: () => ({
```
same comment as in the other snippet regarding this potentially not being needed

yannbf commented on docs/snippets/svelte/msw-addon-configure-handlers-graphql.ts.mdx:
```diff
@@ -43,28 +46,38 @@ const TestData = {
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
This should be awaited
Also don't forget to add `HttpResponse` and `delay` to the msw imports at line 6

yannbf commented on docs/snippets/svelte/msw-addon-configure-handlers-graphql.ts.mdx:
```diff
@@ -10,6 +10,9 @@ import MockApolloWrapperClient from './MockApolloWrapperClient.svelte';
 
 const meta: Meta<typeof DocumentScreen> = {
   component: DocumentScreen,
+  render: () => ({
```
same comment as in other svelte snippets

yannbf commented on docs/snippets/vue/msw-addon-configure-handlers-graphql.ts-4-9.mdx:
```diff
@@ -41,43 +45,40 @@ const TestData = {
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
This should be awaited
Also don't forget to add `HttpResponse` and `delay` to the msw imports at line 4

yannbf commented on docs/snippets/vue/msw-addon-configure-handlers-graphql.ts.mdx:
```diff
@@ -41,43 +45,40 @@ const TestData = {
+    msw: {
+      handlers: [
+        graphql.query('AllInfoQuery', async () => {
+          delay(800);
```
This should be awaited
Also don't forget to add `HttpResponse` and `delay` to the msw imports at line 4

yannbf commented on docs/snippets/web-components/msw-addon-configure-handlers-http.js.mdx:
```diff
@@ -34,21 +34,28 @@ const TestData = {
+    msw: {
+      handlers: [
+        http.get('https://your-restful-endpoint', async () => {
+          await delay(800);
```
Don't forget to add `http`, `HttpResponse` and `delay` to the msw imports at line 4, also remove `rest`

yannbf commented on docs/snippets/web-components/msw-addon-configure-handlers-http.ts.mdx:
```diff
@@ -39,21 +39,28 @@ const TestData = {
+    msw: {
+      handlers: [
+        http.get('https://your-restful-endpoint', async () => {
+          await delay(800);
```
Don't forget to add `http`, `HttpResponse` and `delay` to the msw imports at line 6, also remove `rest`

yannbf reviewed (changes requested):
Awesome work!! I added a few change requests to the msw docs

kylegach committed:
Address MSW snippets feedback

kylegach committed:
Prettify snippets

jonniebigodes commented on docs/writing-stories/mocking-providers.md:
```diff
@@ -0,0 +1,92 @@
+title: Mocking providers
+---
+
+<!-- TODO: React & Solid only? -->
```
Even if this is React and Solid specific, we should at least provide an intro and frame it accordingly wouldn't you agree?

jonniebigodes reviewed (changes requested):
This is shaping up nicely, left a couple of items for you to look into when you have a chance. 

Let me know of any feedback you may have.

kylegach commented on docs/toc.js:
```diff
@@ -141,6 +141,28 @@ module.exports = {
+          title: 'Mocking data and modules',
+          type: 'menu',
+          children: [
+            {
+              pathSegment: 'mocking-providers',
+              title: 'Providers',
+              type: 'link',
+            },
```
I like it!

kylegach replied:
I made it modules, network requests, providers, which is also alphabetical order.

kylegach commented on docs/writing-stories/build-pages-with-storybook.md:
```diff
@@ -73,226 +73,19 @@ This approach is beneficial when the various subcomponents export a complex list
 ## Mocking connected components
 
-If you need to render a connected component in Storybook, you can mock the network requests to fetch its data. There are various layers in which you can do that.
+If you need to render a connected component in Storybook, you can mock the data or modules that component depends on. There are various layers in which you can do that.
```
I took the opportunity to improve this even further:

> Connected components are components that depend on external data or services. For example, a full page component is often a connected component. When you render a connected component in Storybook, you need to mock the data or modules that the component depends on. There are various layers in which you can do that.

kylegach commented on docs/writing-stories/mocking-modules.md:
````diff
@@ -0,0 +1,282 @@
+
+You can then update your component file to use the subpath import:
+
+```ts
````
Good call. I opted for a simpler approach:

````md
```ts
// AuthButton.ts
// ➖ Remove this line
// import { getUserFromSession } from '../../lib/session';
// ➕ Add this line
import { getUserFromSession } from '#lib/session';

// ... rest of the file
```
````

jonniebigodes commented on docs/writing-stories/mocking-network-requests.md:
````diff
@@ -0,0 +1,180 @@
+
+<!-- TODO: Snippetize -->
+
+```sh
+npx storybook@latest add msw-storybook-addon
+```
````
To add some context, this type of addon will not require to be registered in main.js based on its architecture and as it does not provide a UI element. More even, as this one is affecting the preview. Based on some conversations, this will remain as currently documented. Potentially, in the near future, it could be optimized to:
- Be installed via CLI without it being registered
- Automatically inject the required config automatically 🤞

yannbf replied:
Yup the addon is not really an addon in a way, depends on how you see it. It's basically a package that exports a storybook loader. It does not have a preset, it does not have a register file so adding it main.js doesn't really make sense! In the future there are plans to make it into a full on addon with automatic loader, a panel etc, but that's not the case today.

kylegach commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -0,0 +1,282 @@
+
+### Setting up and cleaning up
+
+You can use the asynchronous `beforeEach` function to perform any setup that you need before the story is rendered, eg. setting up mock behavior. It can be defined at the story, component (which will run for all stories in the file), or project (defined in `.storybook/preview.js|ts`, which will run for all stories in the project) level.
```
Great edits. I tweaked it slightly still:

> Before the story renders, you can use the asynchronous `beforeEach` function to perform any setup you need (e.g., configure the mock behavior). This function can be defined at the story, component (which will run for all stories in the file), or project (defined in `.storybook/preview.js|ts`, which will run for all stories in the project).

kylegach committed:
Apply suggestions from code review

Co-authored-by: jonniebigodes <joaocontadesenvolvimento@gmail.com>

kylegach committed:
Address feedback

- Re-order TOC items
- Fix conditional content
- Improve nextjs module API references
- Make example snippets more consistent
- Re-organize subpath imports section of module mocking guide
- Fix typos, grammar

kylegach committed:
Merge branch 'kasper/module-mocking' into module-mocking-docs

kylegach committed:
Snippetize examples

kylegach committed:
Remove unnecessary snippets

kylegach committed:
Prettify snippets

yannbf commented on docs/writing-stories/mocking-providers.md:
```diff
@@ -0,0 +1,88 @@
+
+<Callout variant="info">
+
+The [context provider pattern](https://react.dev/learn/passing-data-deeply-with-context) and how to mock it only applies to renderers that use JSX, like [React](?renderer=react) or [Solid](?renderer=solid).
```
Preact as well?

yannbf commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -86,31 +86,24 @@ To configure subpath imports, you define the `imports` property in your project'
+First, **each subpath must begin with `#`**, to differentiate it from a regular module path. The `#*` entry is a catch-all that maps all subpaths to the root directory.
 
-</Callout>
+Second, note the **`storybook` and `default` keys** in each module's entry. The `storybook` value is used to import the mock file when loaded in Storybook, while the `default` value is used to import the original module in your project. The Storybook environment will match the conditions `storybook` and `test`, so you can apply the same module mapping for both Storybook and your tests.
```
> used to import the original module in your project

Maybe:
used to import the original module **when loaded** in your project ?

yannbf commented on docs/writing-stories/mocking-modules.md:
```diff
@@ -86,31 +86,24 @@ To configure subpath imports, you define the `imports` property in your project'
+There are two aspects to this configuration worth noting:
 
-Each subpath must begin with `#`, to differentiate it from a regular module path. The `#*` entry is a catch-all that maps all subpaths to the root directory.
+First, **each subpath must begin with `#`**, to differentiate it from a regular module path. The `#*` entry is a catch-all that maps all subpaths to the root directory.
```
Might it make sense to mention that this is a standard and point to the docs? https://nodejs.org/api/packages.html#subpath-imports

In case it hasn't been mentioned already, otherwise ignore this comment.

The narrative here is to show people that we are following a standard, not introducing something new for the sake of Storybook

kylegach replied:
The section containing this opens with this sentence:

> The recommended method for mocking modules is to use [subpath imports](https://nodejs.org/api/packages.html#subpath-imports), a feature of Node packages that is supported by both [Vite](../builders/vite.md) and [Webpack](../builders/webpack.md).

yannbf commented on docs/snippets/react/msw-addon-configure-handlers-graphql.js.mdx:
```diff
@@ -63,8 +63,8 @@ export const MockedSuccess = {
             data: {
-              allFilms: {
-                films,
+              AllInfo: {
```
why is this PascalCase?

kylegach replied:
To match the actual query in the component. But perhaps it would be better to update the query to use camelCase instead. I'll do that now.

yannbf commented on docs/api/portable-stories-jest.md:
```diff
@@ -26,10 +26,11 @@ Normally, Storybok composes a story and its [annotations](#annotations) automati
 
 - Configure the [`next/jest.js` transformer](https://nextjs.org/docs/pages/building-your-application/testing/jest#manual-setup), which will handle all of the necessary Next.js configuration for you.
 - Import [`composeStories`](#composestories) or [`composeStory`](#composestory) from the `@storybook/nextjs` package (e.g. `import { composeStories } from '@storybook/nextjs'`).
+- Set up [internal module aliases](../get-started/nextjs.md#storybooknextjsexport-mocks) to be able to mock and assert on them.
```
I think ideally people should always set up the aliases if they can, because right now the aliases only refer to the mocks, but might be extended in the future so that it's required for people to use the aliases. For instance, when we end up supporting multiple Next.js versions where internal paths change based on versions.

yannbf replied:
More info: https://github.com/storybookjs/storybook/pull/26688/files#diff-c1603ee6a674bb75255aa448057630506153e363297fd58668e6679f10909bc6R417

kylegach replied:
Thanks for the info!

yannbf commented on docs/configure/story-rendering.md:
```diff
@@ -2,7 +2,54 @@
+
+<If notRenderer={['angular', 'vue']}>
+
+Here's an example of how you might use the preview file to initialize a library that must run before your components render:
```
Would it make sense to document the usage of `beforeEach` in this docs page?

kylegach replied:
Until we have more canonical documentation for `beforeEach`, I'd rather keep it somewhat limited for now.

kylegach committed:
Address feedback

- Make nextjs module alias usage more clear
- Update msw graphql snippets
- Add preact support for mocking providers
- Prose improvements

kylegach committed:
Update decorator snippets

jonniebigodes reviewed (changes requested):
We're getting there. Left some minor items for you to look into when you have a moment.

kylegach committed:
Address feedback

- Snippet consistency
- Prose tweaks

jonniebigodes commented on docs/snippets/common/before-each-in-meta-mock-date.js.mdx:
````diff
@@ -0,0 +1,26 @@
+```ts
````
Missed one here. when you're able can you adjust this so that it aligns with the file? It shouldn't be `ts` but ´js`, and the same for comment referencing the filename.

jonniebigodes reviewed (approved):
One small item to address and this is good to go. Appreciate you addressing the feedback so promptly 🙏

kylegach committed:
Address feedback

kylegach committed:
Merge branch 'next' into module-mocking-docs
